text/0569-tracked-list.md

```diff
@@ -0,0 +1,306 @@
+
+# TrackedList
+
+## Summary
+
+Introduces a new array-like class whose usage is autotracked by default, and
+is designed with performance in mind.
+
+```js
+import { TrackedList } from '@glimmer/tracking';
+
+class Person {
+  friends = TrackedList.from(['Yehuda', 'Melanie', 'Ricardo']);
+
+  addFriend(newFriend) {
+    this.friends.push(newFriend);
+  }
+}
+```
+
+## Motivation
+
+`@tracked` in Ember Octane was primarily designed around the use case of
+tracking changes to _properties_ on classes. However, there are other forms of
+state that need to be changed in an Ember app, and the most common one is
+tracking changes to _arrays_.
+
+When `@tracked` was introduced, one of the suggested possibilities for tracking
+changes to arrays was the _immutable pattern_:
+
+```js
+class Person {
+  @friends = ['Yehuda', 'Melanie', 'Ricardo'];
+
+  addFriend(newFriend) {
+    this.friends = [...this.friends, newFriend];
+  }
+}
+```
+
+This pattern is solid and works very well for many Octane users, but it is
+fairly opinionated toward a functional-immutable style of programming, one that
+is different from many Ember apps historically, and that conflicts with
+established programming styles in the ecosystem. This is not necessarily a bad
+thing - one of autotracking's strengths is that it is unopinionated about
+programming style, and can be used with many different paradigms - but it means
+that users who prefer an Object Oriented Programming (OOP) style don't have a
+modern, ergonomic equivalent.
+
+We received feedback to this effect from a good number of Octane early adopters,
+which is why we decided to switch back to using `EmberArray` within the guides
+and documentation for our basic examples.
+
+```js
+import { A } from '@ember/array';
+
+class Person {
+  @friends = A(['Yehuda', 'Melanie', 'Ricardo']);
+
+  addFriend(newFriend) {
+    this.friends.pushObject(newFriend);
+  }
+}
+```
+
+`EmberArray` is reactive, so changes to it will update autotracked values and it
+has the behaviors most OOP users want. However, it has a lot of its own baggage:
+
+- KVO methods like `pushObject` and `popObject` have to be learned in order to
+  use the array, which means learning a whole new array API to be effective in
+  Ember.
+- Custom extra methods like `mapBy` and `sortBy` add to the confusion, since
+  they extend the core array API, and sometimes have conflicting names (e.g.
+  `any` vs `some`)
+- Prototype extensions can still be used, and it can be confusing to new users
+  since their DX is fairly nice, even though they have many downsides and are
+  generally recommended against.
+- Performance is not great, since they either rely on patching the array
+  prototype, which is very bad, or on re-implementing most array methods under
+  the hood.
+
+### Introducing `TrackedList`
+
+The goal of `TrackedList` is to provide a better `EmberArray`, in the short
+term, for Ember users who prefer this style of programming; One that lines up
```

Comment:
The RFC should probably mention, if only as an aside, what the long term preference is, if not `TrackedList`.

Reply:
Presumably it's `Proxy`, but it's not totally obvious since later on when `Proxy` is discussed a couple of important caveats are emphasized.

Reply:
IMHO, if this is not the long term solution, it should not be added to core and should remain in an addon.